relayer/lavasession/consumer_session_manager.go

```diff
@@ -75,11 +76,34 @@ func (csm *ConsumerSessionManager) atomicReadCurrentEpoch() (epoch uint64) {
 	return atomic.LoadUint64(&csm.currentEpoch)
 }
 
+// validating we still have providers, otherwise reset valid addresses list
+func (csm *ConsumerSessionManager) validatePairingListNotEmpty() uint64 {
+	csm.lock.RLock() // lock read to validate length
+	numberOfResets := csm.numberOfResets
+	if len(csm.validAddresses) == 0 {
+		csm.lock.RUnlock()                // unlock read
+		csm.lock.Lock()                   // lock write
+		if len(csm.validAddresses) == 0 { // re verify it didn't change while waiting for lock.
+			utils.LavaFormatWarning("Provider pairing list is empty, resetting state.", nil, nil)
+			csm.validAddresses = make([]string, csm.pairingAddressesLength)
+			copy(csm.validAddresses, csm.pairingAddresses) // reset the list of valid addresses
+			csm.numberOfResets += 1
+			numberOfResets = csm.numberOfResets // update numberOfResets with the new value
+		}
+		csm.lock.Unlock() // unlock write
+	} else {
+		csm.lock.RUnlock() // unlock read if length was larger than zero.
+	}
+	return numberOfResets
+}
+
 // GetSession will return a ConsumerSession, given cu needed for that session.
 // The user can also request specific providers to not be included in the search for a session.
 func (csm *ConsumerSessionManager) GetSession(ctx context.Context, cuNeededForSession uint64, initUnwantedProviders map[string]struct{}) (
 	consumerSession *SingleConsumerSession, epoch uint64, providerPublicAddress string, reportedProviders []byte, errRet error,
 ) {
+	numberOfResets := csm.validatePairingListNotEmpty() // if pairing list is empty we reset the state.
```

Comment:
no need to check it on every session, you slow down the main path, do it after an error happens in getting a session, it will run only if we really need this

Reply:
We need to read the number of resets every time, its just a read lock at most cases which doesnt really slow it down as we only write in 2 occasions